ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/OngoingAuthorizationObserver.scala

```diff
@@ -3,14 +3,78 @@
+import com.daml.ledger.api.auth.interceptor.AuthorizationInterceptor
+import com.daml.ledger.participant.state.index.v2.UserManagementStore
+import com.daml.lf.data.Ref
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.server.api.validation.ErrorFactories
+import io.grpc.StatusRuntimeException
 import io.grpc.stub.ServerCallStreamObserver
 
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+/** @param userRightsCheckIntervalInSeconds - determines the interval at which to check whether user rights state has changed.
+  *                                         Also, double of this value serves as timeout value for subsequent user rights state checks.
+  */
 private[auth] final class OngoingAuthorizationObserver[A](
     observer: ServerCallStreamObserver[A],
-    claims: ClaimSet.Claims,
-    authorized: ClaimSet.Claims => Either[AuthorizationError, Unit],
-    throwOnFailure: AuthorizationError => Throwable,
-) extends ServerCallStreamObserver[A] {
+    originalClaims: ClaimSet.Claims,
+    nowF: () => Instant,
+    errorFactories: ErrorFactories,
+    userManagementStore: UserManagementStore,
+    implicit val ec: ExecutionContext,
+    userRightsCheckIntervalInSeconds: Int,
+    akkaScheduler: Scheduler,
+)(implicit loggingContext: LoggingContext)
+    extends ServerCallStreamObserver[A] {
+  self =>
+
+  private val logger = ContextualizedLogger.get(getClass)
+  private val errorLogger = new DamlContextualizedErrorLogger(logger, loggingContext, None)
+
+  @volatile private var lastUserRightsCheckTime = nowF()
+
+  private lazy val userId = originalClaims.applicationId.fold[Ref.UserId](
+    throw new RuntimeException(
+      "Claims were resolved from a user but userId (applicationId) is missing in the claims."
+    )
+  )(Ref.UserId.assertFromString)
+
+  private val cancellableO: Option[Cancellable] = {
+    if (originalClaims.resolvedFromUser) {
+      val delay = userRightsCheckIntervalInSeconds.seconds
+      // Note: https://doc.akka.io/docs/akka/2.6.13/scheduler.html states that:
+      // "All scheduled task will be executed when the ActorSystem is terminated, i.e. the task may execute before its timeout."
+      val c = akkaScheduler.scheduleWithFixedDelay(initialDelay = delay, delay = delay)(runnable =
+        checkUserRights _
+      )
+      Some(c)
+    } else None
+  }
+
+  private def checkUserRights(): Unit = {
+    userManagementStore
+      .listUserRights(userId)
+      .onComplete {
+        case Failure(_) | Success(Left(_)) =>
+          self.synchronized(observer.onError(staleStreamAuthError))
+          cancelUserRightsCheckTask()
+        case Success(Right(userRights)) =>
+          val updatedClaims = AuthorizationInterceptor.convertUserRightsToClaims(userRights)
+          if (updatedClaims.toSet != originalClaims.claims.toSet) {
+            self.synchronized(observer.onError(staleStreamAuthError))
```

Comment:
why do you need to sync for this? looks suspicious :thinking: 
(btw self.synchronized is just synchronized, no? or if you want to make it explicit porbably it would be better to pick the observer for synchronization)
also: if you need that synchronized, probably would be better to wrap the observer into a synchronized observer or something, and not worry about synchronization code in here (hard to prove every call is synchronized)

Reply:
> why do you need to sync for this? 

I synchronized all calls to `onNext`, `onError`, `onComplete` and `setMessageCompression`. Other methods either are documented to be safe to be called from multiplethreads or have to be called early in the lifecycle.
From `io.grpc.stub.ServerCallStreamObserver`:
> Like StreamObserver, implementations are not required to be thread-safe; if multiple threads will be writing to an instance concurrently, the application must synchronize its calls.

Synchronized observer sounds great

Reply:
Changed approach to synchronize methods of `OngoingAuthorizationObserver` rather than just mehtods on the delegate instance